espnet/nets/beam_search_transducer.py

```diff
@@ -0,0 +1,634 @@
+"""Search algorithms for transducer models."""
+
+import numpy as np
+
+import torch
+import torch.nn.functional as F
+
+from espnet.nets.pytorch_backend.transducer.utils import create_lm_batch_state
+from espnet.nets.pytorch_backend.transducer.utils import init_lm_state
+from espnet.nets.pytorch_backend.transducer.utils import is_prefix
+from espnet.nets.pytorch_backend.transducer.utils import recombine_hyps
+from espnet.nets.pytorch_backend.transducer.utils import select_lm_state
+from espnet.nets.pytorch_backend.transducer.utils import substract
+
+
+def greedy_search(decoder, h, recog_args):
+    """Greedy search implementation for transformer-transducer.
+
+    Args:
+        decoder (class): decoder class
+        h (torch.Tensor): encoder hidden state sequences (maxlen_in, Henc)
+        recog_args (Namespace): argument Namespace containing options
+
+    Returns:
+        hyp (list of dicts): 1-best decoding results
+
+    """
+    init_tensor = h.unsqueeze(0)
+    dec_state = decoder.init_state(init_tensor)
+
+    hyp = {
+        "score": 0.0,
```

Comment:
Could you use dataclass or namedtuple instead of dict to represent Hypothesis?

Reply:
https://docs.python.org/3/library/dataclasses.html

Reply:
Sure, I'll take a look!

Reply:
Hum, one question: did you have something in mind when using dataclasses instead of dict? I'm wondering if some refactoring could be done for other algorithms dealing with list of dict originally.

Reply:
Sorry, I'm not sure what you mean. Do you ask if dict can be replaced to dataclasses?

Reply:
This is very basic strategy in Python programming, or generally, prefer style in programming language.
If you use dict object, no one can inspect what attribute exists without accessing it. You can refer some websites about namedtuple and you may know why we use it.

Reply:
Sorry, I guess I badly explained: I understand replacing dict by a dataclass is straightforward in case of greedy decoding. However, for other algorithms we are dealing with list of dict. Instead of simply replacing dict by a dataclass in that cases, I was wondering if something could be done to speed up a bit (e.g.: using list of objects in dataclass and relying on inner methods for sorting, etc, or some other scheme)

Reply:
The beam search class by @ShigekiKarita is good example of using NamedTuple.

Reply:
What code specifically do you have problem with?

Reply:
Basically, dataclasses is just a rich dict, so I don't think speedup or slowdown can be done.

Reply:
I don't have problem to replace dict to dataclass or namedtuple! As I'm not too familiar with these modules I was wondering if I could rethink in a clever way how I deal with hypothesis.

That was just a question I had on top of my head :)

Reply:
There is a slight degradation in terms of decoding speed but I guess it's expected given changes!